### PR TITLE
feat: moving to hover callback

### PR DESCRIPTION
**Pull request by sg-s**
## changes

- [x] red circle on hover using custom hover hack
- [x] hovering should highlight that row in the deeporigin db
- [x] support for reading in JS files using importlib.resources
- [x] fix for bug where first render is a big blue mess
- [x] support for lasso tool
- [x] selecttool that is fed by options from select column
- [x] updating select column should update choices
- [x] lasso tool should write to those rows
- [x] move all JS code to separate files

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -8,6 +8,19 @@ on:
...
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Search for "icecream" in .py files
+        run: |
+          if grep -ri "icecream" --include="*.py" .; then
+            echo "❌🍦: Found 'icecream' in one or more .py files, suggesting that you forgot to remove debug code"
+            exit 1
+          fi
+
```
Comment (sg-s) [inline .github/workflows/main.yml:23]: make sure i don't leave any icecream debug code that can end up in prod

File: MANIFEST.in [hunk 2]
```diff
@@ -1,4 +1,4 @@
 include README.md
 include src/VERSION
 include src/config/default.yml
-include src/data_hub/callback.js
\ No newline at end of file
+include src/data_hub/*.js
```
Comment (sg-s) [inline MANIFEST.in:4]: we now include all JS files in the package

File: src/data_hub/api.py [hunk 3]
```diff
@@ -1125,7 +1125,11 @@ def _row_to_dict(
         elif field.type in ["float", "integer", "boolean"]:
             value = field.value
         elif field.type == "select":
-            value = field.value.selected_options
+            value = field.value
+            if isinstance(value, dict):
+                value = value["selectedOptions"]
+            else:
+                value = value.selected_options
```
Comment (sg-s) [inline src/data_hub/api.py:1132]: this is so strange. for some reason the SDK stoppped sending objects and instead started sending dicts. no idea why

File: src/data_hub/axes_callback.js [hunk 4]
```diff
@@ -11,22 +12,15 @@ const max_size = Math.max(...size_data);
...
 
 // Trigger data change
-source.change.emit();
-
-// callback to update selection in database
-if (typeof window.deeporigin !== "undefined") {
-    var id = source.data['id'][index];
-    deeporigin.dataHub.primaryDatabase.addSelection({ selections: [{ rowId: id }] })
-    
-}
```
Comment (sg-s) [inline src/data_hub/axes_callback.js:32]: shouldn't be here

File: src/data_hub/button_callback.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// Callback when a user presses the button to update labels in the scatter plot Bokeh figure
+
+const selectedData = lasso_selection_source.data;
+const label = label_select.value;
+const labelColumn = label_column_select.value;
+
+console.log("Will write to these rows:", selectedData.ids);
+console.log("Will write this label:", label);
+console.log("Will write to this column:", labelColumn);
+
```
Comment (sg-s) [inline src/data_hub/button_callback.js:10]: i want these in here while we debug on edge

File: src/data_hub/lasso_callback.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+function debounce(func, wait) {
+    let timeout;
+    return function (...args) {
+        const context = this;
+        clearTimeout(timeout);
+        timeout = setTimeout(() => func.apply(context, args), wait);
+    };
+}
+
```
Comment (sg-s) [inline src/data_hub/lasso_callback.js:11]: @kennyjwilli does this make sense? i don't want to hit the API a million times a minute

File: src/data_hub/plotting.py [hunk 7]
```diff
@@ -64,65 +101,212 @@ def scatter(
...
+            options=list(label_data.keys()),
+            width=select_width,
+        )
+
+        label_select = Select(
+            title="Label",
+            value=label_data[first_col][0],
+            options=list(set(label_data[first_col])),
+            width=select_width,
+        )
```
Comment (sg-s) [inline src/data_hub/plotting.py:164]: interactive clustering only supported if there's a select column in the database

File: src/data_hub/api.py (hunk 3 above)
Comment (kennyjwilli) [reply src/data_hub/api.py:1132]: Why does Python even need to make this distinction?

File: MANIFEST.in (hunk 2 above)
Comment (kennyjwilli) [reply MANIFEST.in:4]: Maybe we nest js under it's own dir?

File: src/data_hub/button_callback.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    // Reset selections for all the rows in question
+    const resetChanges = rowIds.map(rowId => ({
+        rowId: rowId,
+        fieldChangeEvents: [{
+            columnId: labelColumn,
+            newValue: { selectedOptions: [] } 
+        }]
+    }));
+
+    deeporigin.dataHub.primaryDatabase.editRows({ changes: resetChanges });
```
Comment (kennyjwilli) [inline src/data_hub/button_callback.js:23]: Hmm I see how you're intending to use this. I think we need to return a promise here and you need to wait for it to complete, else you may get the conflict message.

File: src/data_hub/hover_callback.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+// callback when a user hovers over a point in the scatter plot
+
+if (cb_data.index.indices.length > 0) {
+    // user is actually hovering over a point. 
+    // bokeh has an annoying "feature" where you can hover
+    // over more than 1 point, which we don't want.
+    // we're going to arbitarily choose the first point
+    var chosen_index = (cb_data.index.indices[0]);
```
Comment (kennyjwilli) [inline src/data_hub/hover_callback.js:8]: Prefer `const` when you don't need to change a value.

File: src/data_hub/lasso_callback.js (hunk 6 above)
Comment (kennyjwilli) [reply src/data_hub/lasso_callback.js:11]: Probably, yes. We have lodash available, and you should be able use `_.throttle` ([docs](https://lodash.com/docs/4.17.15#throttle)) if you'd like.

File: src/data_hub/plotting.py [hunk 10]
```diff
@@ -19,43 +24,75 @@ def scatter(
...
     figure_width = 500
     select_width = int(figure_width * 0.3)
-
-    if df.shape[1] < 2:
-        raise ValueError("DataFrame must contain at least two columns.")
+    js_code = _read_js_code()
 
     cols = df.attrs["metadata"]["cols"]
-    cols = [col["name"] for col in cols if col["type"] in ["float", "integer"]]
+    numeric_cols = [col["name"] for col in cols if col["type"] in ["float", "integer"]]
```
Comment (kennyjwilli) [inline src/data_hub/plotting.py:44]: Really feels like `["float", "integer"]` should be pulled out to a top-level def.

**Review by kennyjwilli — APPROVED** [review]
Very cool!

File: src/data_hub/button_callback.js (hunk 8 above)
Comment (sg-s) [reply src/data_hub/button_callback.js:23]: aaaah. i see. ok, let's fix this in the next PR. i want to see if it works at all

File: src/data_hub/plotting.py (hunk 10 above)
Comment (sg-s) [reply src/data_hub/plotting.py:44]: will fix in next PR!

**Pull request merged**